### ISSUE TITLE
Gauntlet 2: Parametric Floor — Why Keep Data?

### ISSUE THREAD

**Issue by samscarrow** [git-scarrow/fast-associative-memory#4]
# Gauntlet 2 — "Parametric Floor" Test (Why Keep Data?)

## Purpose
Red Team Phase 1. Determine whether a single linear layer trained on the streaming data matches or exceeds FAM accuracy. If it does, the entire associative memory mechanism (storing instances) is unnecessary overhead.

## Kill Condition
If a linear probe achieves ≥ FAM top-1 accuracy, the concept of Associative Memory (storing instances) is unnecessary overhead.

## Protocol
- **Backbone:** Same frozen DINOv2 as Gauntlet 1. No fine-tuning. No gradient flow through backbone.
- **Model:** Single linear layer `(D × num_classes)`. No hidden layers, no nonlinearity.
- **Dataset:** Same streaming eval split used for FAM benchmarks. Same class/batch order.
- **Report:** top-1 accuracy, memory (MB), wall-clock training time (seconds).

### 2a — Streaming SGD / AdamW Linear Probe
Standard online linear probe with a small replay buffer.

```python
# Linear layer: nn.Linear(D, num_classes, bias=True)
# Optimizer: AdamW, lr=1e-3, weight_decay=1e-4
# Replay buffer: ring buffer, capacity=256 samples, random replacement
# Each step: sample mini-batch from buffer + current sample, gradient step
# Buffer memory must be << FAM memory budget — document the comparison
```

### 2b — Streaming Ridge Regression (Woodbury / Incremental Closed-Form)
Closed-form solution updated incrementally via Sherman-Morrison rank-1 update. No replay buffer required — O(D²) memory for the covariance matrix.

```python
# Maintain: A = X^T X + lambda * I  (D x D), b = X^T y  (D x num_classes)
# Update A^{-1} each new sample via Sherman-Morrison identity
# Predict: W = A^{-1} b, logits = emb @ W
# lambda is a CLI hyperparameter, default 1e-3
```

## Deliverables
1. `benchmarks/gauntlet_02_linear.py` — standalone script that:
   - Loads frozen DINOv2 embeddings (or computes them if not cached)
   - Runs both 2a (SGD/AdamW probe) and 2b (Streaming Ridge) on the streaming eval set
   - Prints a results table: `method | top-1 accuracy | memory (MB) | train time (s)`
2. Results reproducible with a fixed random seed.
3. CLI args: `--dataset`, `--backbone`, `--seed`, `--buffer-size`, `--lambda`, `--output`.

## Acceptance
Script runs to completion and prints the results table. FAM accuracy (from existing benchmark) included in the table for direct comparison.
